man/pppoe-sniff.8
.\" SPDX-License-Identifier: GPL-2.0-or-later
.TH PPPOE-SNIFF 8 "3 July 2000"
.\""
.UC 4
.SH NAME
pppoe-sniff \- examine network for non-standard PPPoE frames
.SH SYNOPSIS
.B pppoe-sniff \fR[\fIoptions\fR]

.SH DESCRIPTION
\fBpppoe-sniff\fR listens for likely-looking PPPoE PADR and session frames
and deduces extra options required for \fBpppoe(8)\fR to work.

Some DSL providers seem to use non-standard frame types for PPPoE frames,
and/or require a certain value in the Service-Name field.  It is often
easier to sniff those values from a machine which can successfully connect
rather than try to pry them out of the DSL provider.

To use \fBpppoe-sniff\fR, you need two computers, a DSL modem and
an Ethernet hub (\fInot\fR an Ethernet switch.)

If the DSL modem normally connects directly to your computer's
Ethernet card, connect it to the "uplink" port on the Ethernet hub.
Plug two computers into normal ports on the hub.  On one computer, run
whatever software the DSL provider gave you on whatever operating
system the DSL provider supports.  On the other computer, run Linux and
log in as root.

On the Linux machine, put the Ethernet interface into promiscuous mode
and start \fBpppoe-sniff\fR.  If the ethernet interface is \fIeth0\fR,
for example, type these commands:

.nf
	ip link set eth0 promisc on
	pppoe-sniff -I eth0
.fi

On the other machine, start your DSL connection as usual.  After a short
time, \fBpppoe-sniff\fR should print recommendations for the value
of an '-f' option for \fBpppoe\fR.

After \fBpppoe-sniff\fR finishes (or you stop it if it seems hung),
remember to turn off promiscuous mode:

.nf
	ip link set eth0 promisc on
.fi

.SH OPTIONS
.TP
.B \-I \fIinterface\fR
The \fB\-I\fR option specifies the Ethernet interface to use.  Under Linux,
it is typically \fIeth0\fR or \fIeth1\fR.  The interface should be "up"
and in promiscuous mode before you start \fBpppoe-sniff\fR.

.TP
.B \-V
The \fB\-V\fR option causes \fBpppoe-sniff\fR to print its version number and
exit.

.SH BUGS
\fBpppoe-sniff\fR only works on Linux.

.SH AUTHORS
\fBpppoe-sniff\fR was written by Dianne Skoll <dianne@skoll.ca>.

The \fBpppoe\fR home page is \fIhttps://dianne.skoll.ca/projects/rp-pppoe/\fR.

.SH SEE ALSO
pppd(8), pppoe(8), pppoe-server(8), pppoe-relay(8)



doc/CHANGES
# SPDX-License-Identifier: GPL-2.0-or-later

Changes from version 3.14 to 3.15:

- Release 3.15 (2021-05-07)

- src/pppoe.c: Don't ignore SIGTERM and SIGINT.  Send PADT and exit
  if one of those signals is received.

- General: Switch from net-tools (ifconfig and friends) to iproute2 (ip ...)
  on Linux.

Changes from version 3.13 to 3.14:

- Release 3.14 (2020-05-26)

- Delete scripts/pppoe-init-turbolinux.in - TurboLinux is defunct.

- pppoe-server: Add -H and -M options for sending HURL and MOTM packets
  respectively.

- Change VERSION macro to RP_VERSION to avoid conflict with pppd macro.

- Apply all of the patches at
  https://github.com/nabijaczleweli/rp-pppoe/tree/patches/0006-touchups-3.14-BETA-2-v1
  which fix up various compile-time warnings.

- Apply all of the patches at
  https://github.com/nabijaczleweli/rp-pppoe/tree/patches/0004-mostly-cosmetic-3.14-BETA-1-v1
  which implement the following 6 changes:

- 1. Clarify a couple of comments
- 2. Remove the '@' before the command that installs the plugin, in the Makefile
- 3. Fix a typo in doc/HOW-TO-CONNECT
- 4. Use echo instead of /bin/echo in scripts
- 5. Fix potential snprintf(...) truncation
- 6. Cosmetic fix to copyright printout

- Apply all the patches at
  https://github.com/nabijaczleweli/rp-pppoe/tree/patches/0002-pppd-bugfixes-3.13-v1
  which implement the following 13 changes:

- 1. Don't exit if we time out waiting for PADO/PADS packets if
  persist option set
- 2. Remove unused variable
- 3. Don't error out if device is not Ethernet if doit is 0 in PPPoEDevnameHook
- 4. Use one 8-bit bitfield and macros for both ver and type in PPPoEPacket
- 5. Remove unnecessary casts from common.c
- 6. Replace strDup() with POSIX strdup(3)
- 7. Extract expiration check into time_left() function
- 8. Suppress false error message on PPPoE disconnect
- 9. Send PADT on PPPoE disconnect
- 10. Various fixes for errors found by coverity static analysis
- 11. Remove all references to cdefs.h
- 12. Eliminate snprintf-related compiler warning in plugin.c
- 13. Make tag parsing loop condition more accurate

Changes from version 3.12 to 3.13:

- Release 3.13 (2018-11-25)

- Fix potential use-after-free bug

- Properly detecte kernel-mode PPPoE; fix compilation problems on
  Debian Jessie and Debian Stretch.

Changes from version 3.11 to 3.12:

- Update author's name to "Dianne Skoll"

- Add "-W" option to pppoe executable to explicitly set a Host-Unique value

- Fix a few compiler warnings and configure script bugs.

- Removed dead code

Changes from Version 3.10 to 3.11:

- Make the rp-pppoe.so plugin avoid calling exit() if the "persist" pppd
  option was given.

- Permit both PPPoE server and client to specify an MTU/MRU of 1500 assuming
  the underlying Ethernet interface has an MTU of at least 1508.  The larger
  MTU is negotiated per RFC 4638.  NOTE: Only available with kernel-mode
  plugin, not user-mode pty redirector.

- Get rid of hard-coded nobsdcomp, nodeflate, novj and novjccomp options
  in pppoe-server.

- Apply patch from Philip Prindeville to allow cross-compiling.

- Handle UNIT=xxx directive in pppoe-connect. (Keith Waters)

- Add "-i" flag to pppoe-server to ignore PADIs if there are no free sessions.
  Change suggested by Miroslav Lednicky.

- Ignore PADTs not destined for our ethernet address.

- Add "-X <pidfile>" option to pppoe-server.

- Cleanup Makefile.in. (Philip Prindeville)

- In plugin, don't set devnam if it's already set. (Mike Ireton)

- Eliminate race condition in signal handling.

- pppoe-server: Add -q and -Q options to allow specification of path to
  pppd and pppoe programs respectively.

- Add a few more sanity checks to pppoe-server packet handling.

- Use -fno-strict-aliasing when compiling with gcc to avoid possible problems
  from -O2 optimization level.

Changes from Version 3.9 to 3.10: (30 June 2008)

- Fixed compilation problems on various platforms.

- The Makefiles now use (standard) DESTDIR instead of (non-standard)
  RPM_INSTALL_ROOT to relocate installed files.

- Spec file has been updated (it had languished since 3.6.)

Changes from Version 3.8 to 3.9: (21 June 2008)

- pppoe-server has new "-x" option to limit the number of sessions per
  MAC address.

- Added proper timeout handling while waiting for PADO/PADS.

- Fix race condition with some access concentrators that move very quickly
  into session mode (problem noted by Luigi Sgro)

- Fixed compilation problem on BSD.

- Fixed compilation problems with old versions of gcc

- Remove superfluous options in scripts/pppoe-connect.in

Changes from Version 3.7 to 3.8: (2 April 2006)

- Adjusted code and made it possible to disable debugging code to shrink
  size of pppoe executable.

- Fixed bug in MD5 code that caused pppoe-server to segfault on 64-bit
  machines.

- Made various functions and variables static that didn't need to be visible
  outside their source files.

Changes from Version 3.6 to 3.7:

- Fixed typo in the firewall-standalone sample firewall script.
  Fix courtesy of Robert Vogelgesang <vogel@users.sourceforge.net>

- Added -O option to pppoe-server to let you specify a different default
  options file for pppd instead of /etc/ppp/pppoe-server-options
  Feature courtesy of Robert Vogelgesang <vogel@users.sourceforge.net>

- Fixed some silliness and incorrectness in configure.in.
  Feature courtesy of Robert Vogelgesang <vogel@users.sourceforge.net>

- Fixed a typo in pppoe-connect.in that made it fail if used with the
  kernel-mode plugin.

- Make pppoe-server prepend "nic-" to interface name if used with
  kernel-mode plugin.  This lets you use interfaces that don't start
  with "eth" more easily.


Changes from Version 3.5 to 3.6:

- Changed the names of commands from adsl-* to pppoe-* to more logically
  name the scripts.  NOTE INCOMPATIBILITY:

  OLD NAME                   NEW NAME
  adsl-start                 pppoe-start
  adsl-stop                  pppoe-stop
  adsl-status                pppoe-status

src/pppoe-sniff.c
/***********************************************************************
*
* pppoe-sniff.c
*
* Sniff a network for likely-looking PPPoE frames and deduce the
* command-line options to add to pppoe.  USE AT YOUR OWN RISK.
*
* Copyright (C) 2000-2018 by Roaring Penguin Software Inc.
* Copyright (C) 2018-2023 Dianne Skoll
*
* This program may be distributed according to the terms of the GNU
* General Public License, version 2 or (at your option) any later version.
*
* SPDX-License-Identifier: GPL-2.0-or-later
*
***********************************************************************/

#define _GNU_SOURCE 1

#include <getopt.h>

#include <errno.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include "pppoe.h"

/* Default interface if no -I option given */
#define DEFAULT_IF "eth0"

/* Global vars */
int SeenPADR = 0;
int SeenSess = 0;
uint16_t SessType, DiscType;

char *IfName = NULL;		/* Interface name */
char *ServiceName = NULL;	/* Service name   */

/**********************************************************************
*%FUNCTION: parsePADRTags
*%ARGUMENTS:
* type -- tag type
* len -- tag length
* data -- tag data
* extra -- extra user data.
*%RETURNS:
* Nothing
*%DESCRIPTION:
* Picks interesting tags out of a PADR packet
***********************************************************************/
void
parsePADRTags(uint16_t type, uint16_t len, unsigned char *data,
	      void *extra)
{
    switch(type) {
    case TAG_SERVICE_NAME:
	ServiceName = malloc(len+1);
	if (ServiceName) {
	    memcpy(ServiceName, data, len);
	    ServiceName[len] = 0;
	}
	break;
    }
}

/**********************************************************************
*%FUNCTION: fatalSys
*%ARGUMENTS:
* str -- error message
*%RETURNS:
* Nothing
*%DESCRIPTION:
* Prints a message plus the errno value to stderr and exits.
***********************************************************************/
void
fatalSys(char const *str)
{
    printErr("%.256s: %.256s", str, strerror(errno));
    exit(EXIT_FAILURE);
}

/**********************************************************************
*%FUNCTION: rp_fatal
*%ARGUMENTS:
* str -- error message
*%RETURNS:
* Nothing
*%DESCRIPTION:
* Prints a message to stderr and syslog and exits.
***********************************************************************/
void
rp_fatal(char const *str)
{
    printErr("%s", str);
    exit(EXIT_FAILURE);
}

/**********************************************************************
*%FUNCTION: usage
*%ARGUMENTS:
* argv0 -- program name
*%RETURNS:
* Nothing
*%DESCRIPTION:
* Prints usage information and exits.
***********************************************************************/
void
usage(char const *argv0)
{
    fprintf(stderr, "Usage: %s [options]\n", argv0);
    fprintf(stderr, "Options:\n");
    fprintf(stderr, "   -I if_name     -- Specify interface (default %s.)\n",
	    DEFAULT_IF);
    fprintf(stderr, "   -V             -- Print version and exit.\n");
    fprintf(stderr, "\nPPPoE Version %s, Copyright (C) 2000 Roaring Penguin Software Inc.\n", RP_VERSION);
    fprintf(stderr, "              %*s  Copyright (C) 2018-2023 Dianne Skoll\n", (int) strlen(RP_VERSION), "");
    fprintf(stderr, "PPPoE comes with ABSOLUTELY NO WARRANTY.\n");
    fprintf(stderr, "This is free software, and you are welcome to redistribute it under the terms\n");
    fprintf(stderr, "of the GNU General Public License, version 2 or any later version.\n");
    fprintf(stderr, "https://dianne.skoll.ca/projects/rp-pppoe/\n");
    exit(EXIT_SUCCESS);
}

/**********************************************************************
*%FUNCTION: main
*%ARGUMENTS:
* argc, argv -- count and values of command-line arguments
*%RETURNS:
* Nothing
*%DESCRIPTION:
* Main program
***********************************************************************/
int
main(int argc, char *argv[])
{
    int opt;
    int sock;
    PPPoEPacket pkt;
    int size;

    if (getuid() != geteuid() ||
	getgid() != getegid()) {
	fprintf(stderr, "SECURITY WARNING: pppoe-sniff will NOT run suid or sgid.  Fix your installation.\n");
	exit(EXIT_FAILURE);
    }

    while((opt = getopt(argc, argv, "I:V")) != -1) {
	switch(opt) {
	case 'I':
	    SET_STRING(IfName, optarg);
	    break;
	case 'V':
	    printf("pppoe-sniff: RP-PPPoE Version %s\n", RP_VERSION);
	    exit(EXIT_SUCCESS);
	default:
	    usage(argv[0]);
	}
    }

    /* Pick a default interface name */
    if (!IfName) {
	IfName = DEFAULT_IF;
    }

    sock = openInterface(IfName, ETH_P_ALL,  NULL, NULL);

    /* We assume interface is in promiscuous mode -- use "ip link
       show" to ensure this */
    fprintf(stderr, "Sniffing for PADR.  Start your connection on another machine...\n");
    while (!SeenPADR) {
	if (receivePacket(sock, &pkt, &size) < 0) continue;
	if (ntohs(pkt.length) + HDR_SIZE > size) continue;
	if (PPPOE_VER(pkt.vertype) != 1 || PPPOE_TYPE(pkt.vertype) != 1) continue;
	if (pkt.code != CODE_PADR)               continue;

	/* Looks promising... parse it */
	if (parsePacket(&pkt, parsePADRTags, NULL) < 0) {
	    continue;
	}
	DiscType = ntohs(pkt.ethHdr.h_proto);
	fprintf(stderr, "\nExcellent!  Sniffed a likely-looking PADR.\n");
	break;
    }

    while (!SeenSess) {
	if (receivePacket(sock, &pkt, &size) < 0) continue;
	if (ntohs(pkt.length) + HDR_SIZE > size) continue;
	if (PPPOE_VER(pkt.vertype) != 1 || PPPOE_TYPE(pkt.vertype) != 1) continue;
	if (pkt.code != CODE_SESS)               continue;

	/* Cool! */
	SessType = ntohs(pkt.ethHdr.h_proto);
	break;
    }

    fprintf(stderr, "Wonderful!  Sniffed a likely-looking session packet.\n");
    if ((ServiceName == NULL || *ServiceName == 0) &&
	DiscType == ETH_PPPOE_DISCOVERY &&
	SessType == ETH_PPPOE_SESSION) {
	fprintf(stderr, "\nGreat!  It looks like a standard PPPoE service.\nYou should not need any special command-line options.\n");
	return 0;
    }

    fprintf(stderr, "\nOK, looks like you need extra arguments for 'pppoe'.\n");
    if (ServiceName != NULL && *ServiceName != 0) {
	fprintf(stderr, "-S '%s'\n", ServiceName);
    }
    if (DiscType != ETH_PPPOE_DISCOVERY || SessType != ETH_PPPOE_SESSION) {
	fprintf(stderr, "-f %x:%x\n", DiscType, SessType);
    }
    return 0;
}

/**********************************************************************
*%FUNCTION: sysErr
*%ARGUMENTS:
* str -- error message
*%RETURNS:
* Nothing
*%DESCRIPTION:
* Prints a message plus the errno value to syslog.
***********************************************************************/
void
sysErr(char const *str)
{
    printErr("%.256s: %.256s", str, strerror(errno));
}
